### ISSUE TITLE
Reproduce with limited GPU memory

### ISSUE THREAD

**Issue by dltkddn0525** [manhhabui/mDSDI#1]
Hi, I'm trying to reproduce your result but it is hard to run the code directly because of GPU memory issue.

Since your source code does not support multi-GPU training, I simply wrapped each models(`zi_model,zs_model,classifier,domain_discriminator,zs_domain_classifier`) with `nn.DataParallel` after [this line](https://github.com/manhhabui/mDSDI/blob/main/algorithms/mDSDI/src/Trainer_mDSDI.py#L191
) and applied mixed precision with `torch.cuda.amp.GradScaler()` in the training code. 

Because of the intrinsic memory imbalance problem in `nn.DataParallel`, I could only run your code with 2 GPUs by using half of the batch size(16 per domain) in PACS dataset(I've tested on TITAN-X(12GB) and V100(16GB)). However, I got quite lower performance than the reported one.
When target domain is Art, I got 75.73% of test accuracy but the reported performance is 87.7% +- 0.4(I referred to the Apendix in your paper).

Is there any result with smaller batch size that I could refer to? Or do you have any other recommendations to reproduce your results with limited GPU memory?

By the way, I think the bottleneck is the computation of covariance matrix(https://github.com/manhhabui/mDSDI/blob/main/algorithms/mDSDI/src/Trainer_mDSDI.py#L325) as you mentioned in your paper, but I couldn't get why the covariance matrix results in the size of (Batch size x feature dimension x feature dimension) rather than just (feature dimension x feature dimension). If this is a bug, I might run your code with full batch size by only using `nn.DataParallel`.

**Comment by dltkddn0525** [comment]
I ran you source code `python main.py --config 'algorithms/mDSDI/configs/PACS_<target domain>.json` without any modification on single GPU, and still could not reproduce your results.. 

I got following results on PACS dataset
Art : 84.52, Cartoon : 78.92, Photo : 94.73, Sketch : 71.62
where the reported performance is
Art: 87.7+-0.4, Cartoon : 80.4+-0.7, Photo : 98.1+-0.3, Sketch : 78.4 +- 1.2

Your domainbed implementation also gave similar results, which are still below the reported one. Is there anything I missed?

**Comment by manhhabui** [comment]
Hi @dltkddn0525 Suho Lee: thanks for your interest in our work and sorry for the late reply, I just recognized this issue since you emailed me. 
- About the covariance matrix question, we minimized the loss of the whole batch, therefore, it must be (batch size x latent dim x latent dim).
- About reproduce result: I am seeing your two results vary across GPUs and I wonder if this is an issue to not reproduce the result. We used: “NVIDIA DGX-1 Tesla A100-SXM4-40GB”, could you please disclose which GPU you are using for training? We will check.

**Comment by manhhabui** [comment]
Also about the covariance matrix question, I’m not sure I understand what you mean. 

I am trying to explain: In each instance in the batch, we decompose two latents, then compute the covariance between them (latent_size, latent_size). Therefore, for the whole batch, we got (batch_size, latent_size, latent_size). 

We grant this is a computational issue, since it is very large and depends on batch_size as well as latent_size (which we also mentioned in our paper that this is our limitation). 

If you have any idea how to reduce (latent_size, latent_size)? Feel free to share and we are very welcome to improve our paper in the next version.

Thanks.

**Comment by dltkddn0525** [comment]
Thanks for kind replies.

1. The latest results I got(Art : 84.52, Cartoon : 78.92, Photo : 94.73, Sketch : 71.62) are reproduced by using one "A100-SXM-80GB" GPU without any modification of the source code. I wonder if the hyperparameters to reproduce the result are different from the default setting(configurations) of the code.

2. From my understanding, you're trying to make the covariance of every dimension-pair of two random vector(Z_I,Z_S in your paper) equal to zero. As far as I know, the covariance between i-th dimension of random vector X and j-th dimension of random vector Y is calculated as follows:

![cov(X_i,Y_j)&space;=&space;E[(X_i-E[X_i])(Y_j-E[Y_j])]](https://latex.codecogs.com/svg.image?cov(X_i,Y_j)&space;=&space;E[(X_i-E[X_i])(Y_j-E[Y_j])])
If E[X<sub>i</sub>]=E[Y<sub>j</sub>]=0 for all i,j(zero-meaned),
![E[(X_i-E[X_i])(Y_j-E[Y_j])]=E[X_iY_j]=\frac{1}{N}\sum_{k=1}^{N}X_i^k&space;Y_j^k](https://latex.codecogs.com/svg.image?E[(X_i-E[X_i])(Y_j-E[Y_j])]=E[X_iY_j]=\frac{1}{N}\sum_{k=1}^{N}X_i^k&space;Y_j^k)
where N is batch size and X_i^k indicates X_i of k-th sample.

This can be implemented as `cov = torch.mm(zi.T,zs) / batch_size` when each dimension of Z_I, Z_S are zero-meaned, where`cov[i,j]` indicates covariance between i-th dimension of Z_I and j-th dimension of Z_S 

Taking a summation along the batch axis and dividing by batch size from your implementation(i.e., `cov = torch.sum(zi.unsqueeze(2) * zs.unsqueeze(1),dim=0) / batch_size`) gives the same matrix, but can `zi.unsqueeze(2) * zs.unsqueeze(1)` be interpreted as covariance matrix? Is there any specific reason that covariance loss should be computed for each sample in a batch? If not, I think using `cov = torch.mm(zi.T,zs) / batch_size` will reduce the GPU memory issue significantly and also it is more appropriate in terms of "making covariance to zero". Please let me know if there's anything wrong in my answer.

**Comment by manhhabui** [comment]
1. We checked the hyperparams and couldn’t see any difference with what was reported in the paper. I think only these things may be different: GPU type, random seed (since we do not fix seed), environment? I will open this issue for others.
2. Yes, you are correct. This is our bug. We forgot to divide the batch_size of C after https://github.com/manhhabui/mDSDI/blob/main/algorithms/mDSDI/src/Trainer_mDSDI.py#L325 (i.e., `C = torch.mean(C, 0)` must be added in the line 326, this will also reduce the GPU’s storage of the covariance loss). We need to revise and fix this code. Thank you so much for your constructive feedback!

**Comment by dltkddn0525** [comment]
Thanks. Is it possible for you to share the reproduced result after revision? I don't think the difference in GPU type or environment will have a significant impact on performance.

**Comment by manhhabui** [comment]
Okay, I will try to run and share them soon.

**Comment by manhhabui** [comment]
And also, if you have run the result after editing covariance loss, feel free to share your result. Thanks.

**Comment by FRESHMAN11392** [comment]
Thanks to the author for the excellent work, is the fixed code released?
